src/NLog/Targets/TargetWithContext.cs

```diff
@@ -276,6 +277,104 @@ protected bool ShouldIncludeProperties(LogEventInfo logEvent)
+                bool checkExcludeProperties = ExcludeProperties.Count > 0;
+
+                Action<string, object> tryAddProperty = (string propertyName, object propertyValue) =>
+                {
+                    if (string.IsNullOrEmpty(propertyName))
+                        return;
+
+                    if (checkExcludeProperties && ExcludeProperties.Contains(propertyName))
+                        return;
+
+                    if (SerializeScopeContextProperty(logEvent, propertyName, propertyValue, out var serializedValue))
+                    {
+                        combinedProperties.Add(new KeyValuePair<string, object>(propertyName, propertyValue));
+                    }
+                };
+
+                if(logEventProperties?.Count > 0)
+                {
+                    foreach (var property in logEventProperties)
+                        tryAddProperty(property.Key.ToString(), property.Value);
+                }
+
+                if (gdcNames?.Count > 0)
+                {
+                    foreach (var propertyName in gdcNames)
+                        tryAddProperty(propertyName, GlobalDiagnosticsContext.GetObject(propertyName));
+                }
+
+                if (scopeContextProperties?.Any() == true)
+                {
+                    using (var scopeEnumerator = new ScopeContextPropertyEnumerator<object>(scopeContextProperties))
+                    {
+                        while (scopeEnumerator.MoveNext())
+                        {
+                            var scopeProperty = scopeEnumerator.Current;
+                            tryAddProperty(scopeProperty.Key, scopeProperty.Value);
+                        }
+                    }
+                }
+
+                return combinedProperties;
+            }
+
+            return Enumerable.Empty<KeyValuePair<string, object>>();
+        }
+
+        private int EstimateAllPropertiesCount(LogEventInfo logEvent,
+            out IDictionary<object, object> logEventProperties,
+            out IEnumerable<KeyValuePair<string, object>> scopeContextProperties,
+            out ICollection<string> gdcNames,
+            out int propertiesCount)
+        {
+            propertiesCount = ContextProperties?.Count ?? 0;
+
+            if (IncludeEventProperties && logEvent.HasProperties)
+            {
+                logEventProperties = logEvent.Properties;
+                propertiesCount += logEventProperties.Count;
+            }
+            else
+            {
+                logEventProperties = null;
+            }
+
+            if (IncludeScopeProperties && !logEvent.TryGetCachedLayoutValue(_contextLayout.ScopeContextPropertiesLayout, out object value))
+            {
+                scopeContextProperties = ScopeContext.GetAllProperties();
+                propertiesCount += scopeContextProperties.Count();
+            }
+            else
+            {
+                scopeContextProperties = null;
```

Comment:
Believe you are missing logic from `CombineProperties` about casting `out object value` into `IDictionary` and extracting the properties.

Reply:
I think I don't have to destructure the CombineProperties() method, just call it with reference to null:
(new code)

```
if (_targetContext.IncludeScopeProperties)
{
    IDictionary<string, object> combinedProperties = null;
    var contextLayout = _targetContext._contextLayout.ScopeContextPropertiesLayout;
    if (!_targetContext.CombineProperties(_logEvent, contextLayout, ref combinedProperties))
    {
        _scopeContextProperties = ScopeContext.GetAllProperties();
        _propertiesCount += _scopeContextProperties.Count();
    }
}

```